src/main/scala/is/hail/expr/AST.scala

```diff
@@ -406,12 +398,42 @@ case class StructConstructor(posn: Position, names: Array[String], elements: Arr
     celements <- CM.sequence(elements.map(_.compile()))
   ) yield arrayToAnnotation(CompilationHelp.arrayOf(celements))
 
+  def toIR(agg: Option[String] = None): Option[IR]
+}
+
+case class StructConstructor(posn: Position, names: Array[String], elements: Array[AST]) extends BaseStructConstructor(posn, elements) {
+  override def typecheckThis(ec: EvalContext): Type = {
+    elements.foreach(_.typecheck(ec))
+    val types = elements.map(_.`type`)
```

Comment:
no need to produce nice errors: can just be `elements.map(_.'type'.asInstanceOf[Type])`